// @farcaster/snap/dist/react-native/v1/snap-view.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { useEffect, useState } from "react";
import { View, Text, StyleSheet, Pressable } from "react-native";
import { SnapThemeProvider, useSnapTheme } from "../theme.js";
import { SnapLoadingOverlay, SnapViewCoreInner, resolveAccentHex, } from "../snap-view-core.js";
const SNAP_MAX_HEIGHT = 500;
// ─── SnapViewV1 (no validation) ──────────────────────
export function SnapViewV1Inner({ snap, handlers, loading = false, loadingOverlay, }) {
    return (_jsx(SnapViewCoreInner, { snap: snap, handlers: handlers, loading: loading, loadingOverlay: loadingOverlay }));
}
export function SnapViewV1({ snap, handlers, loading = false, appearance = "dark", colors, loadingOverlay, }) {
    return (_jsx(SnapThemeProvider, { appearance: appearance, colors: colors, children: _jsx(SnapViewV1Inner, { snap: snap, handlers: handlers, loading: loading, loadingOverlay: loadingOverlay }) }));
}
// ─── SnapCardV1 (card frame with expandable clipping) ──
function SnapCardV1Inner({ snap, handlers, loading = false, borderRadius, actionError, appearance, plain, loadingOverlay, }) {
    const { colors, mode } = useSnapTheme();
    const accentHex = resolveAccentHex(snap.theme?.accent, mode);
    const [contentHeight, setContentHeight] = useState(0);
    const [isExpanded, setIsExpanded] = useState(false);
    useEffect(() => {
        setIsExpanded(false);
        setContentHeight(0);
    }, [snap]);
    const isExpandable = contentHeight > SNAP_MAX_HEIGHT + 1;
    const isClipped = isExpandable && !isExpanded;
    const isDark = mode === "dark";
    const pillBg = isDark ? "rgba(40,40,40,0.92)" : "rgba(255,255,255,0.92)";
    const pillBgPressed = isDark ? "rgba(60,60,60,0.95)" : "rgba(240,240,240,0.95)";
    return (_jsxs(_Fragment, { children: [_jsxs(View, { style: cardStyles.frameRing, children: [_jsxs(View, { style: [
                            plain ? undefined : cardStyles.card,
                            plain ? undefined : {
                                borderRadius,
                                borderColor: colors.border,
                                backgroundColor: colors.surface,
                            },
                        ], children: [_jsx(View, { style: isClipped ? { maxHeight: SNAP_MAX_HEIGHT, overflow: "hidden" } : undefined, children: _jsx(View, { collapsable: false, onLayout: (event) => {
                                        const nextHeight = Math.round(event.nativeEvent.layout.height);
                                        setContentHeight((currentHeight) => isClipped
                                            ? Math.max(currentHeight, nextHeight)
                                            : currentHeight === nextHeight
                                                ? currentHeight
                                                : nextHeight);
                                    }, style: plain ? undefined : cardStyles.body, children: _jsx(SnapViewV1Inner, { snap: snap, handlers: handlers, loading: loading, loadingOverlay: null }) }) }), loading
                                ? loadingOverlay === undefined
                                    ? _jsx(SnapLoadingOverlay, { appearance: mode, accentHex: accentHex })
                                    : loadingOverlay
                                : null] }), isExpandable ? (_jsx(View, { pointerEvents: "box-none", style: cardStyles.expandFloat, children: _jsx(Pressable, { style: ({ pressed }) => [
                                cardStyles.expandButton,
                                {
                                    backgroundColor: pressed ? pillBgPressed : pillBg,
                                    borderColor: colors.border,
                                },
                            ], onPress: () => {
                                setIsExpanded((value) => !value);
                            }, children: _jsx(Text, { style: [cardStyles.expandButtonText, { color: colors.text }], children: isExpanded ? "Show less" : "Show more" }) }) })) : null] }), actionError && (_jsx(Text, { style: [
                    cardStyles.actionError,
                    {
                        color: appearance === "dark"
                            ? "rgba(255,100,100,0.9)"
                            : "rgba(200,0,0,0.8)",
                    },
                ], children: actionError }))] }));
}
export function SnapCardV1({ snap, handlers, loading = false, appearance = "dark", colors, borderRadius = 16, actionError, plain = false, loadingOverlay, }) {
    return (_jsx(SnapThemeProvider, { appearance: appearance, colors: colors, children: _jsx(SnapCardV1Inner, { snap: snap, handlers: handlers, loading: loading, borderRadius: borderRadius, actionError: actionError, appearance: appearance, plain: plain, loadingOverlay: loadingOverlay }) }));
}
const cardStyles = StyleSheet.create({
    frameRing: { alignSelf: "stretch" },
    card: { overflow: "hidden", borderWidth: 1, minHeight: 120 },
    body: { paddingHorizontal: 16, paddingVertical: 16 },
    expandFloat: {
        position: "absolute",
        left: 0,
        right: 0,
        bottom: -14,
        height: 28,
        alignItems: "center",
        justifyContent: "center",
    },
    expandButton: {
        minWidth: 92,
        alignItems: "center",
        justifyContent: "center",
        borderRadius: 9999,
        borderWidth: 1,
        paddingHorizontal: 10,
        paddingVertical: 4,
    },
    expandButtonText: {
        fontSize: 12,
        lineHeight: 16,
        fontWeight: "600",
    },
    actionError: { paddingHorizontal: 12, paddingVertical: 8, fontSize: 13 },
});
