// @farcaster/snap/dist/react-native/components/snap-action-button.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Pressable, StyleSheet, Text, View } from "react-native";
import { ExternalLink } from "lucide-react-native";
import { useSnapPalette } from "../use-snap-palette.js";
import { useSnapTheme } from "../theme.js";
import { ICON_MAP } from "./snap-icon.js";
function isExternalLinkAction(on) {
    if (!on)
        return false;
    const press = on.press;
    if (!press)
        return false;
    return press.action === "open_url";
}
export function SnapActionButton({ element, emit, }) {
    const { accentHex } = useSnapPalette();
    const { colors } = useSnapTheme();
    const { props } = element;
    const label = String(props.label ?? "Action");
    const variant = String(props.variant ?? "secondary");
    const isPrimary = variant === "primary";
    const iconName = props.icon ? String(props.icon) : undefined;
    const textColor = isPrimary ? "#fff" : colors.text;
    const iconColor = isPrimary ? "#fff" : colors.text;
    const on = element.on;
    const showExternalIcon = isExternalLinkAction(on);
    return (_jsx(View, { style: styles.outer, children: _jsxs(Pressable, { style: ({ pressed }) => [
                styles.btn,
                isPrimary ? styles.btnDefault : styles.btnOther,
                isPrimary
                    ? { backgroundColor: pressed ? accentHex + "DD" : accentHex }
                    : { backgroundColor: pressed ? colors.mutedHover : colors.muted },
                pressed && styles.pressed,
            ], onPress: () => {
                void (async () => {
                    try {
                        await emit("press");
                    }
                    catch (err) {
                        if (typeof __DEV__ !== "undefined" && __DEV__) {
                            console.error("[snap] action failed", err);
                        }
                    }
                })();
            }, children: [iconName && ICON_MAP[iconName]
                    ? (() => {
                        const I = ICON_MAP[iconName];
                        return _jsx(I, { size: 16, color: iconColor });
                    })()
                    : null, _jsx(Text, { style: { color: textColor, fontSize: 14, lineHeight: 18, fontWeight: "600" }, children: label }), showExternalIcon ? (_jsx(ExternalLink, { size: 14, color: iconColor, style: { opacity: 0.6 } })) : null] }) }));
}
const styles = StyleSheet.create({
    outer: { minWidth: 0 },
    btn: {
        paddingHorizontal: 16,
        borderRadius: 10,
        alignItems: "center",
        justifyContent: "center",
        flexDirection: "row",
        gap: 8,
    },
    btnDefault: {
        paddingVertical: 10,
    },
    btnOther: {
        paddingVertical: 8,
    },
    pressed: { opacity: 0.88 },
});
