// @farcaster/snap/dist/react-native/components/snap-slider.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useStateStore } from "@json-render/react-native";
import Slider from "@react-native-community/slider";
import { StyleSheet, Text, View } from "react-native";
import { useSnapPalette } from "../use-snap-palette.js";
import { useSnapTheme } from "../theme.js";
export function SnapSlider({ element: { props }, }) {
    const { get, set } = useStateStore();
    const { accentHex } = useSnapPalette();
    const { colors } = useSnapTheme();
    const name = String(props.name ?? "slider");
    const path = `/inputs/${name}`;
    const min = Number(props.min ?? 0);
    const max = Number(props.max ?? 100);
    const step = props.step != null ? Number(props.step) : 1;
    const fallback = props.defaultValue != null ? Number(props.defaultValue) : (min + max) / 2;
    const raw = get(path);
    const value = raw === undefined || raw === null ? fallback : Number(raw);
    const clamped = Number.isFinite(value)
        ? Math.min(max, Math.max(min, value))
        : fallback;
    const label = props.label != null ? String(props.label) : null;
    const showValue = props.showValue === true;
    const minLabel = props.minLabel != null ? String(props.minLabel) : null;
    const maxLabel = props.maxLabel != null ? String(props.maxLabel) : null;
    return (_jsxs(View, { style: styles.wrap, children: [label ? (_jsxs(View, { style: styles.labelRow, children: [_jsx(Text, { style: [styles.label, { color: colors.text }], children: label }), showValue && (_jsx(Text, { style: [styles.valueText, { color: colors.textSecondary }], children: String(Math.round(clamped)) }))] })) : null, _jsx(Slider, { style: styles.slider, minimumValue: min, maximumValue: max, step: step > 0 ? step : 1, value: clamped, onValueChange: (v) => set(path, v), minimumTrackTintColor: accentHex, maximumTrackTintColor: colors.muted, thumbTintColor: accentHex }), minLabel != null || maxLabel != null ? (_jsxs(View, { style: styles.minMaxRow, children: [_jsx(Text, { style: [styles.minMax, { color: colors.textSecondary }], children: minLabel ?? String(min) }), _jsx(Text, { style: [styles.minMax, { color: colors.textSecondary }], children: maxLabel ?? String(max) })] })) : null] }));
}
const styles = StyleSheet.create({
    wrap: { width: "100%", gap: 2 },
    labelRow: {
        flexDirection: "row",
        justifyContent: "space-between",
        alignItems: "center",
    },
    label: { fontSize: 13, lineHeight: 18, fontWeight: "500", flex: 1 },
    valueText: { fontSize: 13, lineHeight: 18 },
    slider: { width: "100%", height: 40 },
    minMaxRow: {
        flexDirection: "row",
        justifyContent: "space-between",
    },
    minMax: { fontSize: 12, lineHeight: 16 },
});
