// @farcaster/snap/dist/react-native/components/snap-item.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { StyleSheet, Text, View } from "react-native";
import { useSnapStackDirection } from "../stack-direction-context.js";
import { useSnapTheme } from "../theme.js";
export function SnapItem({ element: { props }, children, }) {
    const { colors } = useSnapTheme();
    const title = String(props.title ?? "");
    const description = props.description
        ? String(props.description)
        : undefined;
    /** Match web `Item className="flex-1"`: row peers must share width or title/description collapse. */
    const rowPeer = useSnapStackDirection() === "horizontal";
    const containerVariant = { paddingVertical: 6, paddingHorizontal: 10 };
    return (_jsxs(View, { style: [
            styles.container,
            containerVariant,
            rowPeer && styles.rowPeer,
        ], children: [_jsxs(View, { style: styles.content, children: [title ? _jsx(Text, { style: [styles.title, { color: colors.text }], children: title }) : null, description ? (_jsx(Text, { style: [styles.description, { color: colors.textSecondary }], children: description })) : null] }), children ? (_jsx(View, { style: styles.actions, children: _jsx(View, { style: { flex: 0 }, children: children }) })) : null] }));
}
const styles = StyleSheet.create({
    container: {
        flexDirection: "row",
        alignItems: "center",
    },
    rowPeer: {
        flex: 1,
        minWidth: 0,
    },
    content: {
        flex: 1,
    },
    title: {
        fontSize: 15,
        lineHeight: 20,
        fontWeight: "500",
    },
    description: {
        fontSize: 13,
        lineHeight: 18,
        marginTop: 1,
    },
    actions: {
        marginLeft: "auto",
        paddingLeft: 12,
        flexDirection: "row",
        alignItems: "center",
        flexShrink: 0,
        flexGrow: 0,
        flexBasis: "auto",
        gap: 4,
    },
});
