// @farcaster/snap/dist/react-native/components/snap-bar-chart.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { StyleSheet, Text, View } from "react-native";
import { useSnapPalette } from "../use-snap-palette.js";
import { useSnapTheme } from "../theme.js";
export function SnapBarChart({ element: { props }, }) {
    const { accentHex, hex } = useSnapPalette();
    const { colors } = useSnapTheme();
    const bars = Array.isArray(props.bars) ? props.bars : [];
    const chartColor = String(props.color ?? "accent");
    const maxVal = props.max != null
        ? Number(props.max)
        : Math.max(...bars.map((b) => Number(b.value ?? 0)), 1);
    function barFill(bar) {
        if (bar.color)
            return hex(bar.color);
        if (chartColor !== "accent")
            return hex(chartColor);
        return accentHex;
    }
    return (_jsx(View, { style: styles.wrap, children: bars.map((bar, i) => {
            const value = Number(bar.value ?? 0);
            const pct = maxVal > 0 ? Math.min(100, (value / maxVal) * 100) : 0;
            return (_jsxs(View, { style: styles.row, children: [_jsx(Text, { style: [styles.label, { color: colors.textSecondary }], numberOfLines: 1, children: String(bar.label ?? "") }), _jsx(View, { style: [styles.track, { backgroundColor: colors.muted }], children: _jsx(View, { style: [
                                styles.fill,
                                {
                                    width: `${pct}%`,
                                    backgroundColor: barFill(bar),
                                },
                            ] }) }), _jsx(Text, { style: [styles.value, { color: colors.textSecondary }], children: value })] }, i));
        }) }));
}
const styles = StyleSheet.create({
    wrap: { width: "100%", gap: 8 },
    row: { flexDirection: "row", alignItems: "center", gap: 8 },
    label: { width: 80, fontSize: 12, lineHeight: 16, textAlign: "right" },
    track: { flex: 1, height: 10, borderRadius: 9999, overflow: "hidden" },
    fill: { height: "100%", borderRadius: 9999 },
    value: { width: 32, fontSize: 12, lineHeight: 16, fontVariant: ["tabular-nums"] },
});
