// @farcaster/snap/dist/react-native/components/snap-text.js
import { jsx as _jsx } from "react/jsx-runtime";
import { StyleSheet, Text, View } from "react-native";
import { useSnapStackDirection } from "../stack-direction-context.js";
import { useSnapTheme } from "../theme.js";
const SIZE_STYLES = {
    md: { fontSize: 16, lineHeight: 24 },
    sm: { fontSize: 13, lineHeight: 18 },
};
const WEIGHT_MAP = {
    bold: "700",
    normal: "400",
};
export function SnapText({ element: { props }, }) {
    const { colors } = useSnapTheme();
    const content = String(props.content ?? "");
    const size = String(props.size ?? "md");
    const weight = props.weight ? String(props.weight) : undefined;
    const align = props.align ?? undefined;
    const sizeStyle = SIZE_STYLES[size] ?? SIZE_STYLES.md;
    const resolvedWeight = weight ? WEIGHT_MAP[weight] : sizeStyle?.fontWeight;
    const textAlign = align === "center" ? "center" : align === "right" ? "right" : "left";
    const inHorizontalStack = useSnapStackDirection() === "horizontal";
    return (_jsx(View, { style: inHorizontalStack ? styles.wrapRow : styles.wrapCol, children: _jsx(Text, { style: [
                styles.base,
                {
                    color: colors.text,
                    fontSize: sizeStyle.fontSize,
                    lineHeight: sizeStyle.lineHeight,
                    fontWeight: resolvedWeight,
                    textAlign,
                },
            ], children: content }) }));
}
const styles = StyleSheet.create({
    /** Full width for vertical stacks (alignment / wrapping). */
    wrapCol: { width: "100%" },
    /** Row peers: hug content; avoid width 100% fighting nowrap horizontal rows. */
    wrapRow: { flexShrink: 1, minWidth: 0 },
    base: {},
});
