// @farcaster/snap/dist/react-native/components/snap-toggle-group.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useStateStore } from "@json-render/react-native";
import { Pressable, StyleSheet, Text, View } from "react-native";
import { useSnapTheme } from "../theme.js";
export function SnapToggleGroup({ element: { props }, }) {
    const { get, set } = useStateStore();
    const { colors } = useSnapTheme();
    const name = String(props.name ?? "toggle_group");
    const path = `/inputs/${name}`;
    const label = props.label ? String(props.label) : undefined;
    const isMultiple = Boolean(props.multiple);
    const orientation = String(props.orientation ?? "horizontal");
    const options = Array.isArray(props.options)
        ? props.options
        : [];
    const raw = get(path);
    const defaultValue = props.defaultValue;
    const selected = (() => {
        if (raw !== undefined && raw !== null) {
            return isMultiple
                ? Array.isArray(raw)
                    ? raw
                    : []
                : typeof raw === "string"
                    ? [raw]
                    : [];
        }
        if (defaultValue !== undefined) {
            return Array.isArray(defaultValue)
                ? defaultValue
                : [String(defaultValue)];
        }
        return [];
    })();
    const isVertical = orientation === "vertical";
    const handlePress = (opt) => {
        if (isMultiple) {
            const next = selected.includes(opt)
                ? selected.filter((s) => s !== opt)
                : [...selected, opt];
            set(path, next);
        }
        else {
            if (opt && opt !== selected[0])
                set(path, opt);
        }
    };
    return (_jsxs(View, { style: styles.wrap, children: [label ? _jsx(Text, { style: [styles.label, { color: colors.text }], children: label }) : null, _jsx(View, { style: [
                    styles.group,
                    { backgroundColor: colors.muted },
                    isVertical ? styles.groupVertical : styles.groupHorizontal,
                ], children: options.map((opt, index) => {
                    const isSelected = selected.includes(opt);
                    return (_jsx(Pressable, { style: ({ pressed }) => [
                            styles.option,
                            {
                                backgroundColor: isSelected
                                    ? colors.mutedSelected
                                    : pressed
                                        ? colors.mutedHover
                                        : colors.mutedSubtle,
                            },
                            !isVertical && styles.optionHorizontal,
                        ], onPress: () => handlePress(opt), children: _jsx(Text, { style: [
                                styles.optionText,
                                { color: colors.text },
                            ], children: opt }) }, index));
                }) })] }));
}
const styles = StyleSheet.create({
    wrap: { width: "100%", gap: 6 },
    label: { fontSize: 13, lineHeight: 18, fontWeight: "500" },
    group: {
        padding: 4,
        borderRadius: 8,
        gap: 4,
    },
    groupHorizontal: {
        flexDirection: "row",
    },
    groupVertical: {
        flexDirection: "column",
    },
    option: {
        paddingVertical: 8,
        paddingHorizontal: 12,
        borderRadius: 6,
        alignItems: "center",
        justifyContent: "center",
    },
    optionHorizontal: {
        flex: 1,
    },
    optionText: {
        fontSize: 13,
        lineHeight: 18,
        fontWeight: "500",
    },
});
