// @farcaster/snap/dist/react/hooks/use-snap-colors.js
"use client";
import { useMemo } from "react";
import { useStateStore } from "@json-render/react";
import { resolveSnapPaletteHex } from "../lib/resolve-palette-hex.js";
import { useSnapPreviewPageAccent, useSnapAppearance } from "../accent-context.js";
import { resolveSnapColorHex } from "@farcaster/snap";
/** Readable foreground color (black or white) for a given hex background. */
export function pickForegroundForBg(hex) {
    const h = hex.replace(/^#/, "");
    if (h.length !== 6)
        return "#ffffff";
    const r = Number.parseInt(h.slice(0, 2), 16);
    const g = Number.parseInt(h.slice(2, 4), 16);
    const b = Number.parseInt(h.slice(4, 6), 16);
    const yiq = (r * 299 + g * 587 + b * 114) / 1000;
    return yiq >= 180 ? "#0a0a0a" : "#ffffff";
}
const NEUTRAL_LIGHT = {
    text: "#111111",
    textMuted: "#6B7280",
    border: "#E5E7EB",
    muted: "rgba(0,0,0,0.06)",
    surface: "#ffffff",
    inputBorder: "#E5E7EB",
    inputBg: "rgba(0,0,0,0.06)",
};
const NEUTRAL_DARK = {
    text: "#FAFAFA",
    textMuted: "#A1A1AA",
    border: "#2D2D44",
    muted: "rgba(255,255,255,0.03)",
    surface: "#23262f",
    inputBorder: "#3F3F46",
    inputBg: "rgba(255,255,255,0.03)",
};
function buildSnapColors(accentName, mode) {
    const accent = resolveSnapPaletteHex(accentName, mode);
    const accentFg = pickForegroundForBg(accent);
    const neutrals = mode === "dark" ? NEUTRAL_DARK : NEUTRAL_LIGHT;
    const accentHover = mode === "light"
        ? `color-mix(in srgb, ${accent} 82%, #000000)`
        : `color-mix(in srgb, ${accent} 78%, #ffffff)`;
    const outlineHover = `color-mix(in srgb, ${accent} 14%, ${neutrals.surface})`;
    const paletteHex = (name) => resolveSnapPaletteHex(name, mode);
    const colorHex = (name) => resolveSnapColorHex(name, { accentHex: accent, appearance: mode });
    return {
        accent,
        accentFg,
        accentHover,
        outlineHover,
        ...neutrals,
        mode,
        paletteHex,
        colorHex,
    };
}
/**
 * Returns fully resolved color values for snap components.
 * All colors are concrete hex values (or color-mix expressions for hover states)
 * so they can be used as inline styles, independent of host app CSS.
 */
export function useSnapColors() {
    const { get } = useStateStore();
    const mode = useSnapAppearance();
    const pageAccent = useSnapPreviewPageAccent();
    const fromState = get("/theme/accent");
    const accentRaw = (typeof pageAccent === "string" && pageAccent.length > 0
        ? pageAccent
        : fromState) ?? undefined;
    const accentName = typeof accentRaw === "string" && accentRaw.length > 0
        ? accentRaw
        : "purple";
    return useMemo(() => buildSnapColors(accentName, mode), [accentName, mode]);
}
