// @tursodatabase/serverless@0.2.4/dist/compat.js
import { Session } from './session.js';
import { AsyncLock } from './async-lock.js';
import { DatabaseError } from './error.js';
/**
 * libSQL-compatible error class with error codes.
 */
export class LibsqlError extends Error {
    constructor(message, code, extendedCode, rawCode, cause) {
        super(message);
        this.name = 'LibsqlError';
        this.code = code;
        this.extendedCode = extendedCode;
        this.rawCode = rawCode;
        this.cause = cause;
    }
}
class LibSQLClient {
    constructor(config) {
        this.execLock = new AsyncLock();
        this._closed = false;
        this._defaultSafeIntegers = false;
        this.validateConfig(config);
        const sessionConfig = {
            url: config.url,
            authToken: config.authToken || '',
            remoteEncryptionKey: config.remoteEncryptionKey
        };
        this.session = new Session(sessionConfig);
    }
    validateConfig(config) {
        // Check for unsupported config options
        const unsupportedOptions = [];
        if (config.encryptionKey !== undefined) {
            unsupportedOptions.push({ key: 'encryptionKey', value: config.encryptionKey });
        }
        if (config.syncUrl !== undefined) {
            unsupportedOptions.push({ key: 'syncUrl', value: config.syncUrl });
        }
        if (config.syncInterval !== undefined) {
            unsupportedOptions.push({ key: 'syncInterval', value: config.syncInterval });
        }
        if (config.readYourWrites !== undefined) {
            unsupportedOptions.push({ key: 'readYourWrites', value: config.readYourWrites });
        }
        if (config.offline !== undefined) {
            unsupportedOptions.push({ key: 'offline', value: config.offline });
        }
        if (config.tls !== undefined) {
            unsupportedOptions.push({ key: 'tls', value: config.tls });
        }
        if (config.intMode !== undefined) {
            unsupportedOptions.push({ key: 'intMode', value: config.intMode });
        }
        if (config.fetch !== undefined) {
            unsupportedOptions.push({ key: 'fetch', value: config.fetch });
        }
        if (config.concurrency !== undefined) {
            unsupportedOptions.push({ key: 'concurrency', value: config.concurrency });
        }
        if (unsupportedOptions.length > 0) {
            const optionsList = unsupportedOptions.map(opt => `'${opt.key}'`).join(', ');
            throw new LibsqlError(`Unsupported configuration options: ${optionsList}. Only 'url', 'authToken', and 'remoteEncryptionKey' are supported in the serverless compatibility layer.`, "UNSUPPORTED_CONFIG");
        }
        // Validate required options
        if (!config.url) {
            throw new LibsqlError("Missing required 'url' configuration option", "MISSING_URL");
        }
    }
    get closed() {
        return this._closed;
    }
    get protocol() {
        return "http";
    }
    normalizeStatement(stmt) {
        if (typeof stmt === 'string') {
            return { sql: stmt, args: [] };
        }
        const args = stmt.args || [];
        if (Array.isArray(args)) {
            return { sql: stmt.sql, args };
        }
        // Convert named args to positional args (simplified)
        return { sql: stmt.sql, args: Object.values(args) };
    }
    convertResult(result) {
        const resultSet = {
            columns: result.columns || [],
            columnTypes: result.columnTypes || [],
            rows: result.rows || [],
            rowsAffected: result.rowsAffected || 0,
            lastInsertRowid: result.lastInsertRowid ? BigInt(result.lastInsertRowid) : undefined,
            toJSON() {
                return {
                    columns: this.columns,
                    columnTypes: this.columnTypes,
                    rows: this.rows,
                    rowsAffected: this.rowsAffected,
                    lastInsertRowid: this.lastInsertRowid?.toString()
                };
            }
        };
        return resultSet;
    }
    async execute(stmtOrSql, args) {
        await this.execLock.acquire();
        try {
            if (this._closed) {
                throw new LibsqlError("Client is closed", "CLIENT_CLOSED");
            }
            let normalizedStmt;
            if (typeof stmtOrSql === 'string') {
                const normalizedArgs = args ? (Array.isArray(args) ? args : Object.values(args)) : [];
                normalizedStmt = { sql: stmtOrSql, args: normalizedArgs };
            }
            else {
                normalizedStmt = this.normalizeStatement(stmtOrSql);
            }
            const result = await this.session.execute(normalizedStmt.sql, normalizedStmt.args, this._defaultSafeIntegers);
            return this.convertResult(result);
        }
        catch (error) {
            if (error instanceof LibsqlError) {
                throw error;
            }
            throw mapDatabaseError(error, "EXECUTE_ERROR");
        }
        finally {
            this.execLock.release();
        }
    }
    async batch(stmts, mode) {
        await this.execLock.acquire();
        try {
            if (this._closed) {
                throw new LibsqlError("Client is closed", "CLIENT_CLOSED");
            }
            const sqlStatements = stmts.map(stmt => {
                const normalized = this.normalizeStatement(stmt);
                return normalized.sql; // For now, ignore args in batch
            });
            const result = await this.session.batch(sqlStatements);
            // Return array of result sets (simplified - actual implementation would be more complex)
            return [this.convertResult(result)];
        }
        catch (error) {
            if (error instanceof LibsqlError) {
                throw error;
            }
            throw mapDatabaseError(error, "BATCH_ERROR");
        }
        finally {
            this.execLock.release();
        }
    }
    async migrate(stmts) {
        // For now, just call batch - in a real implementation this would disable foreign keys
        return this.batch(stmts, "write");
    }
    async transaction(mode) {
        throw new LibsqlError("Transactions not implemented", "NOT_IMPLEMENTED");
    }
    async executeMultiple(sql) {
        await this.execLock.acquire();
        try {
            if (this._closed) {
                throw new LibsqlError("Client is closed", "CLIENT_CLOSED");
            }
            await this.session.sequence(sql);
        }
        catch (error) {
            if (error instanceof LibsqlError) {
                throw error;
            }
            throw mapDatabaseError(error, "EXECUTE_MULTIPLE_ERROR");
        }
        finally {
            this.execLock.release();
        }
    }
    async sync() {
        throw new LibsqlError("Sync not supported for remote databases", "NOT_SUPPORTED");
    }
    close() {
        this._closed = true;
        // Note: The libSQL client interface expects synchronous close,
        // but our underlying session needs async close. We'll fire and forget.
        this.session.close().catch(error => {
            console.error('Error closing session:', error);
        });
    }
}
/**
 * Create a libSQL-compatible client for Turso database access.
 *
 * This function provides compatibility with the standard libSQL client API
 * while using the Turso serverless driver under the hood.
 *
 * @param config - Configuration object (only url and authToken are supported)
 * @returns A Client instance compatible with libSQL API
 * @throws LibsqlError if unsupported configuration options are provided
 *
 * @example
 * ```typescript
 * import { createClient } from "@tursodatabase/serverless/compat";
 *
 * const client = createClient({
 *   url: process.env.TURSO_DATABASE_URL,
 *   authToken: process.env.TURSO_AUTH_TOKEN
 * });
 *
 * const result = await client.execute("SELECT * FROM users");
 * console.log(result.rows);
 * ```
 */
export function createClient(config) {
    return new LibSQLClient(config);
}
// Known SQLite base error code names, used to split extended codes like
// "SQLITE_CONSTRAINT_PRIMARYKEY" into base ("SQLITE_CONSTRAINT") and extended.
const sqliteBaseErrorCodes = new Set([
    "SQLITE_ERROR",
    "SQLITE_INTERNAL",
    "SQLITE_PERM",
    "SQLITE_ABORT",
    "SQLITE_BUSY",
    "SQLITE_LOCKED",
    "SQLITE_NOMEM",
    "SQLITE_READONLY",
    "SQLITE_INTERRUPT",
    "SQLITE_IOERR",
    "SQLITE_CORRUPT",
    "SQLITE_NOTFOUND",
    "SQLITE_FULL",
    "SQLITE_CANTOPEN",
    "SQLITE_PROTOCOL",
    "SQLITE_EMPTY",
    "SQLITE_SCHEMA",
    "SQLITE_TOOBIG",
    "SQLITE_CONSTRAINT",
    "SQLITE_MISMATCH",
    "SQLITE_MISUSE",
    "SQLITE_NOLFS",
    "SQLITE_AUTH",
    "SQLITE_FORMAT",
    "SQLITE_RANGE",
    "SQLITE_NOTADB",
    "SQLITE_NOTICE",
    "SQLITE_WARNING",
]);
/**
 * Parse a protocol error code into base and extended codes.
 *
 * The server may send either a base code ("SQLITE_CONSTRAINT") or an extended
 * code ("SQLITE_CONSTRAINT_PRIMARYKEY"). This function splits them so that
 * `code` is always the base code and `extendedCode` carries the full detail.
 */
function parseErrorCode(serverCode) {
    if (sqliteBaseErrorCodes.has(serverCode)) {
        return { code: serverCode };
    }
    // Try to find a base code prefix (e.g. "SQLITE_CONSTRAINT" in "SQLITE_CONSTRAINT_PRIMARYKEY")
    for (const base of sqliteBaseErrorCodes) {
        if (serverCode.startsWith(base + "_")) {
            return { code: base, extendedCode: serverCode };
        }
    }
    // Unknown code — return as-is
    return { code: serverCode };
}
function mapDatabaseError(error, fallbackCode) {
    if (error instanceof DatabaseError && error.code) {
        const { code, extendedCode } = parseErrorCode(error.code);
        return new LibsqlError(error.message, code, extendedCode, error.rawCode, error);
    }
    const cause = error instanceof Error ? error : undefined;
    return new LibsqlError(error.message ?? String(error), fallbackCode, undefined, undefined, cause);
}
