// @tursodatabase/serverless@0.2.4/dist/protocol.js
import { DatabaseError } from './error.js';
export function encodeValue(value) {
    if (value === null || value === undefined) {
        return { type: 'null' };
    }
    if (typeof value === 'number') {
        if (!Number.isFinite(value)) {
            throw new Error("Only finite numbers (not Infinity or NaN) can be passed as arguments");
        }
        return { type: 'float', value };
    }
    if (typeof value === 'bigint') {
        return { type: 'integer', value: value.toString() };
    }
    if (typeof value === 'boolean') {
        return { type: 'integer', value: value ? '1' : '0' };
    }
    if (typeof value === 'string') {
        return { type: 'text', value };
    }
    if (value instanceof ArrayBuffer || value instanceof Uint8Array) {
        const base64 = btoa(String.fromCharCode(...new Uint8Array(value)));
        return { type: 'blob', base64 };
    }
    return { type: 'text', value: String(value) };
}
export function decodeValue(value, safeIntegers = false) {
    switch (value.type) {
        case 'null':
            return null;
        case 'integer':
            if (safeIntegers) {
                return BigInt(value.value);
            }
            return parseInt(value.value, 10);
        case 'float':
            return value.value;
        case 'text':
            return value.value;
        case 'blob':
            if (value.base64) {
                const binaryString = atob(value.base64);
                const bytes = new Uint8Array(binaryString.length);
                for (let i = 0; i < binaryString.length; i++) {
                    bytes[i] = binaryString.charCodeAt(i);
                }
                return Buffer.from(bytes);
            }
            return null;
        default:
            return null;
    }
}
/** HTTP header key for the encryption key */
export const ENCRYPTION_KEY_HEADER = 'x-turso-encryption-key';
export async function executeCursor(url, authToken, request, remoteEncryptionKey) {
    const headers = {
        'Content-Type': 'application/json',
    };
    if (authToken) {
        headers['Authorization'] = `Bearer ${authToken}`;
    }
    if (remoteEncryptionKey) {
        headers[ENCRYPTION_KEY_HEADER] = remoteEncryptionKey;
    }
    const response = await fetch(`${url}/v3/cursor`, {
        method: 'POST',
        headers,
        body: JSON.stringify(request),
    });
    if (!response.ok) {
        let errorMessage = `HTTP error! status: ${response.status}`;
        try {
            const errorBody = await response.text();
            const errorData = JSON.parse(errorBody);
            if (errorData.message) {
                errorMessage = errorData.message;
            }
        }
        catch {
            // If we can't parse the error body, use the default HTTP error message
        }
        throw new DatabaseError(errorMessage);
    }
    const reader = response.body?.getReader();
    if (!reader) {
        throw new DatabaseError('No response body');
    }
    const decoder = new TextDecoder();
    let buffer = '';
    let cursorResponse;
    // First, read until we get the cursor response (first line)
    while (!cursorResponse) {
        const { done, value } = await reader.read();
        if (done)
            break;
        buffer += decoder.decode(value, { stream: true });
        const newlineIndex = buffer.indexOf('\n');
        if (newlineIndex !== -1) {
            const line = buffer.slice(0, newlineIndex).trim();
            buffer = buffer.slice(newlineIndex + 1);
            if (line) {
                cursorResponse = JSON.parse(line);
                break;
            }
        }
    }
    if (!cursorResponse) {
        throw new DatabaseError('No cursor response received');
    }
    async function* parseEntries() {
        try {
            // Process any remaining data in the buffer
            let newlineIndex;
            while ((newlineIndex = buffer.indexOf('\n')) !== -1) {
                const line = buffer.slice(0, newlineIndex).trim();
                buffer = buffer.slice(newlineIndex + 1);
                if (line) {
                    yield JSON.parse(line);
                }
            }
            // Continue reading from the stream
            while (true) {
                const { done, value } = await reader.read();
                if (done)
                    break;
                buffer += decoder.decode(value, { stream: true });
                while ((newlineIndex = buffer.indexOf('\n')) !== -1) {
                    const line = buffer.slice(0, newlineIndex).trim();
                    buffer = buffer.slice(newlineIndex + 1);
                    if (line) {
                        yield JSON.parse(line);
                    }
                }
            }
            // Process any remaining data in the buffer
            if (buffer.trim()) {
                yield JSON.parse(buffer.trim());
            }
        }
        finally {
            reader.releaseLock();
        }
    }
    return { response: cursorResponse, entries: parseEntries() };
}
export async function executePipeline(url, authToken, request, remoteEncryptionKey) {
    const headers = {
        'Content-Type': 'application/json',
    };
    if (authToken) {
        headers['Authorization'] = `Bearer ${authToken}`;
    }
    if (remoteEncryptionKey) {
        headers[ENCRYPTION_KEY_HEADER] = remoteEncryptionKey;
    }
    const response = await fetch(`${url}/v3/pipeline`, {
        method: 'POST',
        headers,
        body: JSON.stringify(request),
    });
    if (!response.ok) {
        throw new DatabaseError(`HTTP error! status: ${response.status}`);
    }
    return response.json();
}
