// @farcaster/snap/dist/server/verify.js
import { decode, decodePayload as jfsDecodePayload, encodePayload as jfsEncodePayload, toJsonFarcasterSignature, verify, } from "@farcaster/jfs";
import { hexToBytes } from "viem";
import { DEFAULT_SNAP_HUB_HTTP_BASE_URL, getActiveEd25519SignerKeysFromHubHttp, } from "./hubs.js";
/**
 * Parse a JFS object or string into normalized form (see {@link toJsonFarcasterSignature} / `uncompact`).
 */
export function parseJfs(text) {
    const trimmed = text.trim();
    let jfsFromJson;
    try {
        jfsFromJson = JSON.parse(trimmed);
    }
    catch {
        jfsFromJson = undefined;
    }
    if (jfsFromJson !== undefined && isJfsObject(jfsFromJson)) {
        return { ok: true, jfs: jfsFromJson };
    }
    const jfsFromString = tryUncompactJfsString(trimmed);
    if (jfsFromString) {
        return { ok: true, jfs: jfsFromString };
    }
    return {
        ok: false,
        error: "invalid JFS envelope: must be JSON with header, payload, and signature fields, or a JFS compact string (three dot-separated segments)",
    };
}
export async function verifyJFS(jfs, options = {}) {
    const decoded = tryDecodeJfs(jfs);
    if (!decoded) {
        return {
            valid: false,
            error: new Error("invalid JFS envelope"),
        };
    }
    try {
        await verify({ data: jfs, strict: true, keyTypes: ["app_key"] });
    }
    catch (error) {
        return {
            valid: false,
            error: error instanceof Error ? error : new Error(String(error)),
        };
    }
    return hubVerifyDecodedPayload(decoded, options);
}
/**
 * Verify that the signing key for a JFS payload is an active signer for the FID.
 */
async function hubVerifyDecodedPayload(decoded, options) {
    const { header, payload } = decoded;
    const keys = await getActiveEd25519SignerKeysFromHubHttp(options.hubHttpBaseUrl ?? DEFAULT_SNAP_HUB_HTTP_BASE_URL, header.fid);
    if (!keys.ok) {
        return {
            valid: false,
            error: new Error(keys.message),
        };
    }
    let headerKeyBytes;
    try {
        headerKeyBytes = hexToBytes(header.key);
    }
    catch {
        return {
            valid: false,
            error: new Error("invalid JFS header key encoding"),
        };
    }
    if (headerKeyBytes.length !== 32) {
        return {
            valid: false,
            error: new Error("JFS app_key public key must be 32 bytes"),
        };
    }
    const matched = keys.signers.some((k) => bytesEqual(k.publicKey, headerKeyBytes));
    if (!matched) {
        return {
            valid: false,
            error: new Error("active hub signer list does not include JFS header key"),
        };
    }
    return {
        valid: true,
        data: payload,
        signingUserFid: header.fid,
    };
}
export function decodePayload(payload) {
    return jfsDecodePayload(payload);
}
export function encodePayload(payload) {
    return jfsEncodePayload(payload);
}
/**
 * Normalize a compact JFS string to `{ header, payload, signature }` using
 * `@farcaster/jfs` {@link toJsonFarcasterSignature} (which delegates to `uncompact` for strings).
 * Returns null if the string is malformed.
 */
function tryUncompactJfsString(value) {
    try {
        return toJsonFarcasterSignature(value.trim());
    }
    catch {
        return null;
    }
}
/**
 * Fully decode a JFS envelope or compact string via `@farcaster/jfs` {@link decode}:
 * parsed header object, parsed payload, signature bytes.
 */
function tryDecodeJfs(input) {
    try {
        return decode(input);
    }
    catch {
        return null;
    }
}
function isJfsObject(v) {
    return (v !== null &&
        typeof v === "object" &&
        "header" in v &&
        typeof v.header === "string" &&
        "payload" in v &&
        typeof v.payload === "string" &&
        "signature" in v &&
        typeof v.signature === "string");
}
function bytesEqual(a, b) {
    if (a.length !== b.length)
        return false;
    let diff = 0;
    for (let i = 0; i < a.length; i++)
        diff |= a[i] ^ b[i];
    return diff === 0;
}
