// @farcaster/snap@2.2.0/dist/react/components/action-button.js
"use client";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useState } from "react";
import { ExternalLink } from "lucide-react";
import { Button } from "@neynar/ui/button";
import { cn } from "@neynar/ui/utils";
import { useSnapColors } from "../hooks/use-snap-colors.js";
import { ICON_MAP } from "./icon.js";
function isExternalLinkAction(on) {
    if (!on)
        return false;
    const press = on.press;
    if (!press)
        return false;
    return press.action === "open_url";
}
export function SnapActionButton({ element, emit, }) {
    const { props } = element;
    const label = String(props.label ?? "Action");
    const variant = String(props.variant ?? "secondary");
    const isPrimary = variant === "primary";
    const iconName = props.icon ? String(props.icon) : undefined;
    const colors = useSnapColors();
    const [hovered, setHovered] = useState(false);
    const Icon = iconName ? ICON_MAP[iconName] : undefined;
    const showExternalIcon = isExternalLinkAction(element.on);
    const style = {
        cursor: "pointer",
        ...(isPrimary
            ? {
                backgroundColor: hovered ? colors.accentHover : colors.accent,
                color: colors.accentFg,
                borderColor: "transparent",
            }
            : {
                backgroundColor: hovered
                    ? `color-mix(in srgb, ${colors.accent} 15%, transparent)`
                    : colors.muted,
                color: colors.text,
                borderColor: "transparent",
            }),
    };
    return (_jsx("div", { className: "w-full min-w-0 flex-1", children: _jsxs(Button, { type: "button", variant: isPrimary ? "default" : "secondary", className: cn("w-full gap-2"), style: style, onClick: () => emit("press"), onPointerEnter: () => setHovered(true), onPointerLeave: () => setHovered(false), children: [Icon && _jsx(Icon, { size: 16 }), label, showExternalIcon && (_jsx(ExternalLink, { size: 14, style: { opacity: 0.6 } }))] }) }));
}
