// @farcaster/snap/dist/react-native/catalog-renderer.js
import { createRenderer } from "@json-render/react-native";
import { snapJsonRenderCatalog } from "@farcaster/snap/ui";
import { SnapActionButton } from "./components/snap-action-button.js";
import { SnapBadge } from "./components/snap-badge.js";
import { SnapIcon } from "./components/snap-icon.js";
import { SnapImage } from "./components/snap-image.js";
import { SnapInput } from "./components/snap-input.js";
import { SnapItem } from "./components/snap-item.js";
import { SnapItemGroup } from "./components/snap-item-group.js";
import { SnapProgress } from "./components/snap-progress.js";
import { SnapSeparator } from "./components/snap-separator.js";
import { SnapSlider } from "./components/snap-slider.js";
import { SnapStack } from "./components/snap-stack.js";
import { SnapSwitch } from "./components/snap-switch.js";
import { SnapText } from "./components/snap-text.js";
import { SnapToggleGroup } from "./components/snap-toggle-group.js";
import { SnapBarChart } from "./components/snap-bar-chart.js";
import { SnapCellGrid } from "./components/snap-cell-grid.js";
/**
 * Maps snap json-render catalog types to React Native primitives.
 * Keys match the snap wire-format `type` strings exactly (snake_case).
 */
export const SnapCatalogView = createRenderer(snapJsonRenderCatalog, {
    badge: SnapBadge,
    button: SnapActionButton,
    icon: SnapIcon,
    image: SnapImage,
    input: SnapInput,
    item: SnapItem,
    item_group: SnapItemGroup,
    progress: SnapProgress,
    separator: SnapSeparator,
    slider: SnapSlider,
    stack: SnapStack,
    switch: SnapSwitch,
    text: SnapText,
    toggle_group: SnapToggleGroup,
    bar_chart: SnapBarChart,
    cell_grid: SnapCellGrid,
});
