// @farcaster/snap/dist/react-native/snap-view-core.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { snapJsonRenderCatalog } from "@farcaster/snap/ui";
import { SnapCatalogView } from "./catalog-renderer.js";
import { ConfettiOverlay } from "./confetti-overlay.js";
import { useSnapTheme } from "./theme.js";
import { useCallback, useEffect, useMemo, useRef, useState, } from "react";
import { ActivityIndicator, StyleSheet, View } from "react-native";
import { DEFAULT_THEME_ACCENT, PALETTE_LIGHT_HEX, PALETTE_DARK_HEX, } from "@farcaster/snap";
// ─── Shared helpers ──────────────────────────────────
export function applyStatePaths(model, changes) {
    const entries = Array.isArray(changes)
        ? changes.map((c) => [c.path, c.value])
        : Object.entries(changes);
    for (const [path, value] of entries) {
        const trimmed = path.startsWith("/") ? path : `/${path}`;
        const parts = trimmed.split("/").filter(Boolean);
        if (parts.length < 2)
            continue;
        const [top, ...rest] = parts;
        if (top === "inputs") {
            if (typeof model.inputs !== "object" || model.inputs === null) {
                model.inputs = {};
            }
            const inputs = model.inputs;
            if (rest.length === 1) {
                inputs[rest[0]] = value;
            }
            continue;
        }
        if (top === "theme") {
            if (typeof model.theme !== "object" || model.theme === null) {
                model.theme = {};
            }
            const theme = model.theme;
            if (rest.length === 1) {
                theme[rest[0]] = value;
            }
        }
    }
}
export function resolveAccentHex(accent, appearance) {
    const map = appearance === "dark" ? PALETTE_DARK_HEX : PALETTE_LIGHT_HEX;
    const name = accent && Object.hasOwn(map, accent)
        ? accent
        : DEFAULT_THEME_ACCENT;
    return map[name];
}
// ─── Core rendering component (no validation) ────────
export function SnapViewCoreInner({ snap, handlers, loading = false, loadingOverlay, }) {
    const { mode } = useSnapTheme();
    const spec = snap.ui;
    const accentHex = resolveAccentHex(snap.theme?.accent, mode);
    const initialState = useMemo(() => ({
        ...(spec.state ?? {}),
        inputs: { ...(spec.state?.inputs ?? {}) },
        theme: {
            ...(spec.state?.theme ?? {}),
            ...(snap.theme ? { accent: snap.theme.accent } : {}),
        },
    }), [spec, snap.theme]);
    const stateRef = useRef(initialState);
    useEffect(() => {
        stateRef.current = {
            inputs: {
                ...(initialState.inputs ?? {}),
            },
            theme: {
                ...(initialState.theme ?? {}),
            },
        };
    }, [initialState]);
    useEffect(() => {
        const catalogResult = snapJsonRenderCatalog.validate(spec);
        if (!catalogResult.success) {
            // eslint-disable-next-line no-console
            console.warn("[Snap] catalog validation issues:", catalogResult.error);
        }
    }, [spec]);
    const [pageKey, setPageKey] = useState(0);
    useEffect(() => {
        setPageKey((k) => k + 1);
    }, [spec]);
    const showConfetti = snap.effects?.includes("confetti") ?? false;
    const [confettiKey, setConfettiKey] = useState(0);
    useEffect(() => {
        if (showConfetti)
            setConfettiKey((k) => k + 1);
    }, [showConfetti, snap]);
    const handlersRef = useRef(handlers);
    handlersRef.current = handlers;
    const handleAction = useCallback((name, params) => {
        const inputs = (stateRef.current.inputs ?? {});
        const p = (params ?? {});
        const h = handlersRef.current;
        switch (name) {
            case "submit":
                h.submit(String(p.target ?? ""), inputs);
                break;
            case "open_url":
                h.open_url(String(p.target ?? ""));
                break;
            case "open_snap":
                h.open_snap(String(p.target ?? ""));
                break;
            case "open_mini_app":
                h.open_mini_app(String(p.target ?? ""));
                break;
            case "view_cast":
                h.view_cast({ hash: String(p.hash ?? "") });
                break;
            case "view_profile":
                h.view_profile({ fid: Number(p.fid ?? 0) });
                break;
            case "compose_cast":
                h.compose_cast({
                    text: p.text ? String(p.text) : undefined,
                    channelKey: p.channelKey ? String(p.channelKey) : undefined,
                    embeds: Array.isArray(p.embeds) ? p.embeds : undefined,
                });
                break;
            case "view_token":
                h.view_token({ token: String(p.token ?? "") });
                break;
            case "send_token":
                h.send_token({
                    token: String(p.token ?? ""),
                    amount: p.amount ? String(p.amount) : undefined,
                    recipientFid: p.recipientFid ? Number(p.recipientFid) : undefined,
                    recipientAddress: p.recipientAddress
                        ? String(p.recipientAddress)
                        : undefined,
                });
                break;
            case "swap_token":
                h.swap_token({
                    sellToken: p.sellToken ? String(p.sellToken) : undefined,
                    buyToken: p.buyToken ? String(p.buyToken) : undefined,
                });
                break;
            default:
                break;
        }
    }, []);
    return (_jsxs(View, { style: styles.container, children: [loading
                ? loadingOverlay === undefined
                    ? (_jsx(SnapLoadingOverlay, { appearance: mode, accentHex: accentHex }))
                    : loadingOverlay
                : null, _jsx(SnapCatalogView, { spec: spec, state: initialState, loading: false, onStateChange: (changes) => {
                    applyStatePaths(stateRef.current, changes);
                }, onAction: handleAction }, pageKey), showConfetti && _jsx(ConfettiOverlay, {}, confettiKey)] }));
}
export function SnapLoadingOverlay({ appearance, accentHex, }) {
    return (_jsx(View, { style: [
            styles.overlay,
            {
                backgroundColor: appearance === "dark"
                    ? "rgba(0,0,0,0.1)"
                    : "rgba(255,255,255,0.2)",
            },
        ], children: _jsx(ActivityIndicator, { size: "large", color: accentHex }) }));
}
const styles = StyleSheet.create({
    container: {
        width: "100%",
    },
    overlay: {
        ...StyleSheet.absoluteFillObject,
        alignItems: "center",
        justifyContent: "center",
        zIndex: 10,
    },
});
