// @farcaster/snap/dist/react-native/theme.js
import { jsx as _jsx } from "react/jsx-runtime";
import { createContext, useContext, useMemo } from "react";
const DEFAULT_LIGHT = {
    bg: "#dfe3e8",
    surface: "#ffffff",
    text: "rgba(0,0,0,0.9)",
    textSecondary: "rgba(0,0,0,0.5)",
    border: "rgba(0,0,0,0.1)",
    inputBg: "rgba(0,0,0,0.06)",
    muted: "rgba(0,0,0,0.08)",
    mutedSubtle: "rgba(0,0,0,0.04)",
    mutedHover: "rgba(0,0,0,0.12)",
    mutedSelected: "rgba(0,0,0,0.16)",
};
const DEFAULT_DARK = {
    bg: "#111318",
    surface: "#1a1d24",
    text: "rgba(255,255,255,0.9)",
    textSecondary: "rgba(255,255,255,0.5)",
    border: "rgba(255,255,255,0.1)",
    inputBg: "rgba(255,255,255,0.04)",
    muted: "rgba(255,255,255,0.06)",
    mutedSubtle: "rgba(255,255,255,0.03)",
    mutedHover: "rgba(255,255,255,0.08)",
    mutedSelected: "rgba(255,255,255,0.12)",
};
const SnapThemeContext = createContext({
    mode: "dark",
    colors: DEFAULT_DARK,
});
export function SnapThemeProvider({ appearance, colors, children, }) {
    const value = useMemo(() => {
        const defaults = appearance === "dark" ? DEFAULT_DARK : DEFAULT_LIGHT;
        return {
            mode: appearance,
            colors: colors ? { ...defaults, ...colors } : defaults,
        };
    }, [appearance, colors]);
    return (_jsx(SnapThemeContext.Provider, { value: value, children: children }));
}
export function useSnapTheme() {
    return useContext(SnapThemeContext);
}
