// @farcaster/snap@2.2.0/dist/react-native/components/snap-badge.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { StyleSheet, Text, View } from "react-native";
import { useSnapPalette } from "../use-snap-palette.js";
import { ICON_MAP } from "./snap-icon.js";
export function SnapBadge({ element: { props }, }) {
    const { accentHex, hex } = useSnapPalette();
    const label = String(props.label ?? "");
    const variant = String(props.variant ?? "default");
    const color = props.color ? String(props.color) : undefined;
    const iconName = props.icon ? String(props.icon) : undefined;
    const isAccent = !color || color === "accent";
    const resolvedColor = isAccent ? accentHex : hex(color);
    const isFilled = variant !== "outline";
    const Icon = iconName ? ICON_MAP[iconName] : undefined;
    return (_jsxs(View, { style: [
            styles.badge,
            isFilled
                ? { backgroundColor: resolvedColor + "20", borderColor: "transparent" }
                : { borderColor: resolvedColor },
        ], children: [Icon && (_jsx(Icon, { size: 12, color: resolvedColor })), _jsx(Text, { style: [
                    styles.label,
                    { color: resolvedColor },
                ], children: label })] }));
}
const styles = StyleSheet.create({
    badge: {
        alignSelf: "flex-start",
        flexDirection: "row",
        alignItems: "center",
        gap: 4,
        paddingHorizontal: 8,
        paddingVertical: 2,
        borderRadius: 9999,
        borderWidth: 1,
    },
    label: {
        fontSize: 12,
        lineHeight: 16,
        fontWeight: "500",
    },
});
