// @farcaster/snap@2.2.0/dist/react-native/components/snap-switch.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useStateStore } from "@json-render/react-native";
import { StyleSheet, Switch, Text, View } from "react-native";
import { useSnapPalette } from "../use-snap-palette.js";
import { useSnapTheme } from "../theme.js";
export function SnapSwitch({ element: { props }, }) {
    const { get, set } = useStateStore();
    const { accentHex } = useSnapPalette();
    const { colors } = useSnapTheme();
    const name = String(props.name ?? "switch");
    const path = `/inputs/${name}`;
    const label = props.label ? String(props.label) : undefined;
    const fallback = Boolean(props.defaultChecked ?? false);
    const raw = get(path);
    const checked = raw === undefined || raw === null ? fallback : Boolean(raw);
    return (_jsxs(View, { style: styles.row, children: [label ? _jsx(Text, { style: [styles.label, { color: colors.text }], children: label }) : null, _jsx(Switch, { value: checked, onValueChange: (v) => set(path, v), trackColor: { false: colors.muted, true: accentHex }, thumbColor: "#fff" })] }));
}
const styles = StyleSheet.create({
    row: {
        flexDirection: "row",
        alignItems: "center",
        justifyContent: "space-between",
        gap: 12,
    },
    label: {
        fontSize: 14,
        lineHeight: 18,
        fontWeight: "400",
        flex: 1,
    },
});
