// @farcaster/snap@2.2.0/dist/react/catalog-renderer.js
"use client";
import { createRenderer } from "@json-render/react";
import { snapJsonRenderCatalog } from "@farcaster/snap/ui";
import { SnapActionButton } from "./components/action-button.js";
import { SnapBadge } from "./components/badge.js";
import { SnapIcon } from "./components/icon.js";
import { SnapImage } from "./components/image.js";
import { SnapInput } from "./components/input.js";
import { SnapItem } from "./components/item.js";
import { SnapItemGroup } from "./components/item-group.js";
import { SnapProgress } from "./components/progress.js";
import { SnapSeparator } from "./components/separator.js";
import { SnapSlider } from "./components/slider.js";
import { SnapStack } from "./components/stack.js";
import { SnapSwitch } from "./components/switch.js";
import { SnapText } from "./components/text.js";
import { SnapToggleGroup } from "./components/toggle-group.js";
import { SnapBarChart } from "./components/bar-chart.js";
import { SnapCellGrid } from "./components/cell-grid.js";
/**
 * Maps snap json-render catalog types to React components.
 * Keys match the snap wire-format `type` strings exactly.
 */
export const SnapCatalogView = createRenderer(snapJsonRenderCatalog, {
    badge: SnapBadge,
    button: SnapActionButton,
    icon: SnapIcon,
    image: SnapImage,
    input: SnapInput,
    item: SnapItem,
    item_group: SnapItemGroup,
    progress: SnapProgress,
    separator: SnapSeparator,
    slider: SnapSlider,
    stack: SnapStack,
    switch: SnapSwitch,
    text: SnapText,
    toggle_group: SnapToggleGroup,
    bar_chart: SnapBarChart,
    cell_grid: SnapCellGrid,
});
